### PR TITLE
[Firefox] Hide the viewBookmark buttons for PDFs opened from file:// URLs, since clicking on those buttons throws a SecurityError in that case

### PR DESCRIPTION
**Pull request by Snuffleupagus**
This PR fixes a Firefox specific issue, and also re-factors a bit of code to improve our lint coverage somewhat.

### CODE REVIEW

File: web/viewer.js [hunk 1]
```diff
@@ -1498,10 +1492,26 @@ function webViewerInitialized() {
...
+  // for file:// URLs, hence we hide the buttons since they won't work.
+  if (file && file.lastIndexOf('file:', 0) === 0) {
+    document.getElementById('viewBookmark').classList.add('hidden');
+    document.getElementById('secondaryViewBookmark').classList.add('hidden');
+  }
+//#endif
 //#if CHROME
-//if (file) {
-//  ChromeCom.openPDFFile(file);
-//}
```
Comment (Rob--W) [inline web/viewer.js]: Why did you change this? `ChromeCom` is only for the Chrome extension, so this should only be enabled when the Chrome extension is build, and not when the PDF viewer is loaded without building.

Comment (Rob--W) [inline web/viewer.js]: Never mind, same name was used for Firefox's com object.

Comment (Rob--W) [inline web/viewer.js]: Addon/extension-specific code should be in `//#if` blocks, to make sure that PDF.js (web/viewer.html) works without building.

**Comment by Rob--W** [comment]
The bookmark link is useful to jump to a specific section within the PDF. The following error seems a bug in Firefox, so we should fix it by fixing Firefox rather than hiding the button, IMO:

> Security Error: Content at resource://pdf.js/web/viewer.html may not load or link to file:///tmp/example.pdf#page=1&zoom=auto,-77,841.

(note: `document.URL`, `location.href`, the location bar, etc. all point to `file:///tmp/example.pdf`, so it shouldn't be a problem to change the fragment of the URL)

File: web/viewer.js (hunk 1 above)
Comment (Snuffleupagus) [inline web/viewer.js]: I'm not sure I really understand your concern here. The viewer still works locally (`node make server`), courtesy of [line 1495](https://github.com/mozilla/pdf.js/pull/6480/files#diff-dd42e9b2d2a652b8e312efa567698aa6R1495) to [line 1499](https://github.com/mozilla/pdf.js/pull/6480/files#diff-dd42e9b2d2a652b8e312efa567698aa6R1499) above.

Note that the build specific code is still enclosed in in preprocessor tags. But I'd like to move away from having blocks of code commented out like this, since that prevents linting, which makes it easier for error to creep in.

Comment (Rob--W) [inline web/viewer.js]: `if(true)return;` is not that great either: if there is a static analysis tool that looks for dead code, then this line would trigger it.

If the concern is linting, then we should run the lint tool for all targets to ensure full coverage.

**Comment by Snuffleupagus** [comment]
> The following error seems a bug in Firefox, so we should fix it by fixing Firefox rather than hiding the button, IMO:

As far as I can tell, this is actually by design, which is why I submitted the PR.
It's supposedly possible for the user to relax this restriction, but it seems to be purposely difficult to do so; see http://kb.mozillazine.org/Links_to_local_pages_don't_work#Disabling_the_Security_Check for some context.

**Comment by Rob--W** [comment]
@Snuffleupagus The SecurityError here is caused by `resource://` trying to "navigate" to `file://`, while all other signs (location bar, DOM API, etc.) suggests that this is a `file://` URL. Changing the URL to a local URL while being at a local URL works fine in Firefox.

Both in Chrome and Firefox, changing `location.hash` at `file:` is permitted (in Chrome, changing the path (e.g. `file:///tmp/foo` -> `file:///tmp/bar`) is forbidden, in Firefox that is allowed as well).

**Comment by timvandermeij** [comment]
What is the status of this PR? If there are doubts about the approach for the refactoring commit, we can also choose to leave that one out for now and really only address the bookmark button (and keep the refactoring for a follow-up).

**Comment by timvandermeij** [comment]
What is left to do here? This has been stale for some time now, so we should either decide to fix it up or to close it.
